storage/filesystem/object.go

```diff
@@ -37,8 +38,67 @@ func (s *ObjectStorage) Set(core.Object) (core.Hash, error) {
 }
 
 // Get returns the object with the given hash, by searching for it in
-// the packfile.
+// the packfile and the git object directories.
 func (s *ObjectStorage) Get(h core.Hash) (core.Object, error) {
+	obj, err := s.getFromUnpacked(h)
+	if err == nil {
+		return obj, nil
+	}
+	if err != dotgit.ErrObjfileNotFound {
+		return nil, err
+	}
+
+	obj, err = s.getFromPackfile(h)
+	if err == nil {
+		return obj, nil
+	}
+
+	return nil, err
+}
+
+func (s *ObjectStorage) getFromUnpacked(h core.Hash) (obj core.Object, err error) {
+	fs, path, err := s.dir.Objectfile(h)
+	if err != nil {
+		return nil, err
+	}
+
+	f, err := fs.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		errClose := f.Close()
+		if err == nil {
+			err = errClose
+		}
+	}()
+
+	obj = s.NewObject()
+	objReader, err := objfile.NewReader(f)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		errClose := objReader.Close()
+		if err == nil {
+			err = errClose
+		}
+	}()
+
+	err = objReader.FillObject(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	return obj, nil
+}
+
+func (s *ObjectStorage) getFromPackfile(h core.Hash) (obj core.Object, err error) {
+	if s.index == nil {
+		return nil, dotgit.ErrIdxNotFound
```

Comment:
1.  Change return value to core.ErrObjectNotFound  in `Get()` like following line
https://github.com/src-d/go-git/blob/v4/storage/memory/storage.go#L136

2. Remove L98-L100 in `getFromPackfile()`

Please give me your comment.